presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java

```diff
@@ -629,67 +576,100 @@ private Iterator<LocatedFileStatus> statusFromPrefixes(List<String> prefixes)
                         false,
                         1,
                         BLOCK_SIZE.toBytes(),
-                        object.getLastModified().getTime(),
-                        qualifiedPath(new Path(PATH_SEPARATOR + object.getKey()))))
+                        object.lastModified().toEpochMilli(),
+                        qualifiedPath(new Path(PATH_SEPARATOR + object.key()))))
                 .map(this::createLocatedFileStatus)
                 .iterator();
     }
 
-    private boolean isGlacierObject(S3ObjectSummary object)
+    private boolean isGlacierObject(S3Object object)
     {
-        return GLACIER_STORAGE_CLASSES.contains(object.getStorageClass());
+        return GLACIER_STORAGE_CLASSES.contains(object.storageClassAsString());
     }
 
-    private boolean isHadoopFolderMarker(S3ObjectSummary object)
+    private boolean isHadoopFolderMarker(S3Object object)
     {
-        return (object.getKey().endsWith(DIRECTORY_SUFFIX) && (object.getSize() == 0));
+        return (object.key().endsWith(DIRECTORY_SUFFIX) && (object.size() == 0));
     }
 
-    /**
-     * This exception is for stopping retries for S3 calls that shouldn't be retried.
-     * For example, "Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: Forbidden (Service: Amazon S3; Status Code: 403 ..."
-     */
-    @VisibleForTesting
-    static class UnrecoverableS3OperationException
-            extends IOException
+    private static boolean isDirectory(PrestoS3ObjectMetadata metadata)
     {
-        public UnrecoverableS3OperationException(Path path, Throwable cause)
-        {
-            // append the path info to the message
-            super(format("%s (Path: %s)", cause, path), cause);
+        HeadObjectResponse response = metadata.getObjectResponse();
+        if (response == null) {
+            // If we have no metadata, it could be a directory (prefix-only)
+            return false;
+        }
+
+        String contentType = response.contentType();
+        if (contentType == null) {
+            // No content type, check if it's an empty object with path separator
+            return metadata.isKeyNeedsPathSeparator() && response.contentLength() == 0;
+        }
+
+        // Check if content length is 0 and key needs path separator (empty directory)
+        if (metadata.isKeyNeedsPathSeparator() && response.contentLength() == 0) {
+            return true;
+        }
+        MediaType mediaType;
+        try {
+            mediaType = MediaType.parse(contentType);
+        }
+        catch (IllegalArgumentException e) {
+            log.debug(e, "Failed to parse contentType [%s], assuming not a directory", contentType);
+            return false;
         }
+
+        return mediaType.is(X_DIRECTORY_MEDIA_TYPE) ||
+                (mediaType.is(OCTET_STREAM_MEDIA_TYPE)
+                        && metadata.isKeyNeedsPathSeparator()
+                        && response.contentLength() == 0);
```

Comment:
Just wanted to get an idea, why did we add an if block earlier on line 621? Due to that, this second condition is never going to hit because it will never reach here.